Framework/Algorithms/src/FFT.cpp

```diff
@@ -110,18 +109,201 @@ void FFT::exec() {
+  constexpr bool centerShift = true;
+
+  for (int i = 0; i < ySize; i++) {
+    int j = centerShift ? (ySize / 2 + i) % ySize : i;
+    data[2 * i] = m_inWS->y(iReal)[j]; // even indexes filled with the real part
+    data[2 * i + 1] = isComplex
+                          ? m_inImagWS->y(iImag)[j]
+                          : 0.; // odd indexes filled with the imaginary part
+  }
+
+  double shift = getPhaseShift(
+      m_inWS->x(iReal)); // extra phase to be applied to the transform
+
+  gsl_fft_complex_forward(data.get(), 1, ySize, wavetable, workspace);
+  /* The Fourier transform overwrites array 'data'. Recall that the Fourier
+  * transform is
+  * periodic along the frequency axis. Thus, 'data' takes the same values
+  * when index j runs
+  * from ySize/2 to ySize than if index j would run from -ySize/2 to 0. Thus,
+  * for negative
+  * frequencies running from -ySize/2 to 0, we use the values stored in array
+  * 'data'
+  * for index j running from ySize/2 to ySize.
+  */
+  for (int i = 0; i < ySize; i++) {
+    int j = (ySize / 2 + i + dys) % ySize;
+    m_outWS->mutableX(m_iRe)[i] =
+        df * (-ySize / 2 + i); // zero frequency at i = ySize/2
+    double re = data[2 * j] *
+                dx; // use j from ySize/2 to ySize for negative frequencies
+    double im = data[2 * j + 1] * dx;
+    // shift
+    {
+      double c = cos(m_outWS->x(m_iRe)[i] * shift);
+      double s = sin(m_outWS->x(m_iRe)[i] * shift);
+      double re1 = re * c - im * s;
+      double im1 = re * s + im * c;
+      re = re1;
+      im = im1;
+    }
+    m_outWS->mutableY(m_iRe)[i] = re;                       // real part
+    m_outWS->mutableY(m_iIm)[i] = im;                       // imaginary part
+    m_outWS->mutableY(m_iAbs)[i] = sqrt(re * re + im * im); // modulus
+    if (addPositiveOnly) {
+      m_outWS->dataX(0)[i] = df * i;
+      if (j < ySize / 2) {
+        m_outWS->mutableY(0)[j] = re;                      // real part
+        m_outWS->mutableY(1)[j] = im;                      // imaginary part
+        m_outWS->mutableY(2)[j] = sqrt(re * re + im * im); // modulus
+      } else {
+        m_outWS->mutableY(0)[j] = 0.; // real part
+        m_outWS->mutableY(1)[j] = 0.; // imaginary part
+        m_outWS->mutableY(2)[j] = 0.; // modulus
+      }
+    }
+  }
+  if (xSize == ySize + 1) {
+    m_outWS->mutableX(0)[ySize] = m_outWS->x(0)[ySize - 1] + df;
+    if (addPositiveOnly)
+      m_outWS->mutableX(m_iRe)[ySize] = m_outWS->x(m_iRe)[ySize - 1] + df;
+  }
+
+  gsl_fft_complex_wavetable_free(wavetable);
+  gsl_fft_complex_workspace_free(workspace);
+}
+
+void FFT::transformBackward(boost::shared_array<double> &data,
+                            const bool addPositiveOnly, bool isComplex,
+                            const int iReal, const int iImag, const double df,
+                            const double dx) {
+
+  const int ySize = static_cast<int>(m_inWS->blocksize());
+  const int xSize = static_cast<int>(m_inWS->x(iReal).size());
```

Comment:
There are things that a duplicated in the forward and backward functions. I would move it outside the scope of these methods to avoid that. Maybe put it in the `exec()`, before and after the call to the forward/backward methods.